source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,95 @@ const re2::RE2& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatElements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) {
+  elements_.clear();
+
+  // This call will iterate through the map tree and serialize the key/values as JSON.
+  // If a string value that contains a substitution commands is found, the current
+  // JSON piece and the substitution command will be pushed into the output list.
+  // After that, the iteration will continue until the whole tree is traversed.
+  formatValueToFormatElements(struct_format.fields());
+  elements_.push_back(FormatElement{std::move(buffer_), false});
+  buffer_.clear();
+
+  return std::move(elements_);
+};
+
+void JsonFormatBuilder::formatValueToFormatElements(const ProtobufWkt::Value& value) {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    serializer_.addNull();
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    serializer_.addNumber(value.number_value());
+    break;
+  case ProtobufWkt::Value::kStringValue: {
+    absl::string_view string_format = value.string_value();
+    if (!absl::StrContains(string_format, '%')) {
```

Comment:
Wrong link for the index: https://github.com/envoyproxy/envoy/pull/35545/files#diff-55d03d93118484f891da85b3ae673a01955faa49745104921d5128f5031e906aR446

Reply:
The `SubstitutionFormatParser::parse` will never return an empty list. There are possible cases for the parse method:

1. empty format: return a list that contains a plain string provider which will output empty string.
2. non-emtpy format:
  2.1 return a list that contains plain string providers and other providers.
  2.2 throw exception to reject the configuration if the format is invalid.


The single `%` will hit the case 2.2.

But in case the some one change the behavior of parse in the future, I can do a minor change to guard that. :)

Reply:
Update the if check to determine if the formatter list size is 1 or not. Now, it should be safe anyway, even someone change the behavior of parse in the future (although I think this won't happen).

```
      if (formatters.size() != 1) {
        // 2. Handle the formatter element with multiple or zero providers.
        stringValueToLogLine(formatters, context, info, serializer);
      } else {
        // 3. Handle the formatter element with a single provider and value
        //    type needs to be kept.
        typedValueToLogLine(formatters, context, info, serializer);
      }
```

Reply:
I'm not sure if this change addresses all of @zuercher 's comments, but I'll leave it to him.

Reply:
I misread the parsing code a bit, and it doesn't do what I thought. I think handling an empty formatter list is not a bad thing though. Thanks for adding the guard.